### PR TITLE
Extractable region

### PR DESCRIPTION
**Pull request by dpomianbuff**
New features in this PR:
* Implemented a new way of checking whether a selection can be extracted or not
* Added UTs for Kotlin code

### CODE REVIEW

File: src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt [hunk 1]
```diff
@@ -70,6 +70,15 @@ class EFCandidateFactory {
...
      * Selected region is adjusted by enlarging it to the top-most PsiElement, either start or end
+     *
+     * Example:
+     * 10. if (x == 0) {
+     * 11.  println(x)
+     * 12.  some more statements
+     * 13.  more statements
+     * 14. }
+     *
+     * If the selection has lines 10-13 (included), then it will enlarge the selection to lines 10-14 (included)
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt:81]: Will it work with higher nesting? Something like:
```
     * 8. if (y == 1) {
     * 9. some code
     * 10. if (x == 0) {
     * 11.  println(x)
     * 12.  some more statements
     * 13.  more statements
     * 14. }
     * 15. }
```
     
What will be if user selects lines 8-13?

File: src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt [hunk 2]
```diff
@@ -108,19 +117,21 @@ class EFCandidateFactory {
     private fun getLeftmostPsiElement(lineNumber: Int, editor: Editor, file: PsiFile): PsiElement? {
         // get the PsiElement on the given lineNumber
         var psiElement = file.findElementAt(editor.document.getLineStartOffset(lineNumber)) ?: return null
+        var psiElementEndOffset = file.findElementAt(editor.document.getLineEndOffset(lineNumber)) ?: return null
 
         // if there are multiple sibling PsiElements on the same line, look for the first one
-        while (psiElement.getLineNumber(true) == psiElement.prevSibling?.getLineNumber(true)) {
+        while (psiElement.getLineNumber(false) == psiElement.prevSibling?.getLineNumber(false)) {
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt:123]: Could you please tell what boolean parameter of `psiElement.getLineNumber()` method means? Why did you change it?

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt [hunk 3]
```diff
@@ -64,3 +76,57 @@ fun identifyExtractFunctionSuggestions(input: String): EFSuggestionList {
...
+    if (elements == null || elements.isEmpty()) {
+        return false
+    }
+
+    var allOptionsToExtract = emptyList<ExtractOptions>()
+
+    try {
+        allOptionsToExtract = findAllOptionsToExtract(elements)
+    } catch (e: Exception) {
+    }
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt:108]: Should we handle the exception somehow? Maybe log it.

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt [hunk 4]
```diff
@@ -64,3 +76,57 @@ fun identifyExtractFunctionSuggestions(input: String): EFSuggestionList {
...
+    ): Boolean {
+        return elements.isNotEmpty()
+    }
+
+    var res = false
+    val efKotlinHandler = ExtractKotlinFunctionHandler()
+    if (file !is KtFile) return false
+    try {
+        efKotlinHandler.selectElements(editor, file) { elements, _ ->
+            res = continuation(elements)
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: `continuation` function contains only one line of code and it is used only here, so I don't see much benefits of extracting this code into a separate method.

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt [hunk 5]
```diff
@@ -64,3 +76,57 @@ fun identifyExtractFunctionSuggestions(input: String): EFSuggestionList {
...
+        }
+    } catch (e: Exception) {
+        res = false
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt:125]: Maybe we need to handle the exception, for example, log it?

File: src/test/kotlin/com/intellij/ml/llm/template/EFCandidateFactoryTest.kt [hunk 6]
```diff
@@ -151,4 +152,96 @@ class EFCandidateFactoryTest : LightPlatformCodeInsightTestCase() {
...
+            functionName = "createPartitionMetadata",
+            lineStart = 113,
+            lineEnd = 120
+        )
+        val candidates = EFCandidateFactory().buildCandidates(efs, editor, file).toTypedArray()
+        TestCase.assertEquals(1, candidates.size)
+        TestCase.assertTrue(isCandidateExtractable(candidates.get(0), editor, file))
+    }
+
+    fun `test candidate is not extractable in Java code`() {
```
Comment (onewhl) [inline src/test/kotlin/com/intellij/ml/llm/template/EFCandidateFactoryTest.kt:168]: Should `adjustRegion` method help to resolve such cases?

**Review by onewhl — CHANGES_REQUESTED** [review]
@dpomianbuff cool, thank you! Please, take a look at comments.

File: src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt (hunk 1 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt:81]: Yes, it will "bubble" up to be at the same level, so if 8-13 is selected, then 8-15 will be extracted in this case

File: src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt (hunk 2 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt:123]: True means it will return the line of the start offset, and False means it will return the line of the end offset. I changed it because I need the end offset now.

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt (hunk 3 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt:108]: Sure, we can log it

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt (hunk 4 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: Agreed, I will remove the function

File: src/test/kotlin/com/intellij/ml/llm/template/EFCandidateFactoryTest.kt (hunk 6 above)
Comment (dpomianbuff) [reply src/test/kotlin/com/intellij/ml/llm/template/EFCandidateFactoryTest.kt:168]: Yes, `adjustRegion` will help solve such case. However, there are two types (so far) of candidates: AS_IS and ADJUSTED. The reason for having these two types is so that we can have stats on how many AS_IS candidates can/can not be extracted. This will help with the evaluation because it will show how many of the ChatGPT suggestions were applicable as is vs how many needed adjustment.

In `test candidate is not extractable in Java code`, I'm purposefully testing that the suggestion of ChatGPT is not extractable as is.

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt (hunk 5 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt:125]: Sure, we can log it

File: .github/workflows/gradle-build.yml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+name: Gradle Build
```
Comment (onewhl) [inline .github/workflows/gradle-build.yml]: Could you please rebase this branch on the latest master to exclude these changes?

Comment (dpomianbuff) [reply .github/workflows/gradle-build.yml]: Done. Rebased and also squashed all commits into one

**Review by onewhl — APPROVED** [review]
